### DISCUSSION TITLE
How can I change the text color of the pop up 'Copied!' when the copy code button is clicked?

### DISCUSSION THREAD

**Discussion by aymennasri** [quarto-dev/quarto-cli#11519] [Q&A] [answered]
### Description

This is the text color in light mode:
![image](https://github.com/user-attachments/assets/865bd5a7-2d21-4f75-bc18-519da7788170)
And somehow it changes in dark mode with no apparent scss rules changing it:
![image](https://github.com/user-attachments/assets/a3d260c5-2f8c-44b6-94d8-9d7eb1182e4e)

Using Linux Mint 21.3

`quarto --version`
1.5.55

**Comment by mcanouil** [comment]
Use your favourite browser developer/inspect mode to identify the correct (S)CSS selector, then set it to whatever you want.

You should be able to find many other discussions here and elsewhere about this process.

Note that a screenshot does not replace actual code, it only help illustrate.

**Reply by aymennasri** [reply]
Tried inspecting and there was no styling variables, tried searching through scss variables and rules for Quarto, nothing worked; and yes I'm aware of the fact that the screenshots do not replace actual code, I'm only illustrating the problem in order to get an answer about a variable name or rule that can help me control the copy code button popup text.

**Reply by mcanouil** [reply]
All variables are listed in https://quarto.org/docs/output-formats/html-themes.html

For instance ["code copy"](https://quarto.org/docs/output-formats/html-themes.html#code-copy), but because you did not share actual code, this might not be the option you are looking for.

**Reply by aymennasri** [reply]
Which part of the custom theme scss file do I need to share?

**Reply by jimjam-slam** [reply, marked as answer]
I can't speak to the SCSS side, but on the CSS side, on quarto.org it looks like the tooltip is styled by the CSS property `--bs-tooltip-bg`:

```css
.tooltip {
  --bs-tooltip-color: #fff;
  --bs-tooltip-bg: #000;
}
```

If you're looking to change this in dark mode specifically, you could add something like:

```css
body.quarto-dark .tooltip {
  --bs-tooltip-bg: #fff;
  --bs-tooltip-color: #000;
}
```

(You might need some extra styling to also change the pointy "tip" of the tooltip, too, but I'm hoping it also uses the custom CSS property!)

**Reply by cderv** [reply]
> Which part of the custom theme scss file do I need to share?

This could be an issue with some themes not correctly adapting to dark mode. 

I believe from Bootstrap, tooltip color should be adapted based on `$body-bg` by default. The SCSS variable from boostrap that controls it is `$tooltip-color`, which defines in CSS `var(--bs-tooltip-color)`

But for some dark theme, it will be set to `var(--bs-emphasis-color);` controlled by `$body-emphasis-color`. 
but I wonder if this should not be `$body-emphasis-color-dark` 🤔 
So as you see this is not easy to track. This is not Quarto SCSS but Bootstrap + Bootswatch theme. 

Can you see which theme are you using here ? I would like to see if there an issue to fix. thanks!

Probably tweaking the CSS as shown by @jimjam-slam is a good way without entering SCSS world.

**Reply by aymennasri** [reply]
> > Which part of the custom theme scss file do I need to share?
> 
> This could be an issue with some themes not correctly adapting to dark mode. 
> 
> I believe from Bootstrap, tooltip color should be adapted based on `$body-bg` by default. The SCSS variable from boostrap that controls it is `$tooltip-color`, which defines in CSS `var(--bs-tooltip-color)`
> 
> But for some dark theme, it will be set to `var(--bs-emphasis-color);` controlled by `$body-emphasis-color`. 
> but I wonder if this should not be `$body-emphasis-color-dark` 🤔 
> So as you see this is not easy to track. This is not Quarto SCSS but Bootstrap + Bootswatch theme. 
> 
> Can you see which theme are you using here ? I would like to see if there an issue to fix. thanks!
> 
> Probably tweaking the CSS as shown by @jimjam-slam is a good way without entering SCSS world. 

I actually have my own SCSS theme that I built so based on your advice, should I try to tinker around with the variables you mentionned? I'd love to have it all in the same SCSS file without tweaking the CSS.

**Reply by aymennasri** [reply]
> I can't speak to the SCSS side, but on the CSS side, on quarto.org it looks like the tooltip is styled by the CSS property `--bs-tooltip-bg`:
> 
> ```css
> .tooltip {
>   --bs-tooltip-color: #fff;
>   --bs-tooltip-bg: #000;
> }
> ```
> 
> If you're looking to change this in dark mode specifically, you could add something like:
> 
> ```css
> body.quarto-dark .tooltip {
>   --bs-tooltip-bg: #fff;
>   --bs-tooltip-color: #000;
> }
> ```
> 
> (You might need some extra styling to also change the pointy "tip" of the tooltip, too, but I'm hoping it also uses the custom CSS property!)

Thanks for your answer! The first code block worked inside the SCSS file with no issues.

**Reply by cderv** [reply]
> I actually have my own SCSS theme that I built so based on your advice, should I try to tinker around with the variables you mentionned? I'd love to have it all in the same SCSS file without tweaking the CSS.

Yes please do share your custom theme if you want help on it. Otherwise, you can get inspiration from boostwatch theme. You can see that they do have specific handling for tooltip to get the coloring right
Se for example darkly
https://github.com/quarto-dev/quarto-cli/blob/88819d5b9f092381d5d20a146ac3121dc2435acc/src/resources/formats/html/bootstrap/themes/darkly.scss#L270-L273

If you have a custom them no inheriting from an existing dark theme you may want to do the same. 

> Thanks for your answer! The first code block worked inside the SCSS file with no issues.

But this is basically the CSS shared above that you are happy with. It is just a matter of knowing if you want to reuse a sass variable or use specific color for your theme.

**Reply by aymennasri** [reply]
> > I actually have my own SCSS theme that I built so based on your advice, should I try to tinker around with the variables you mentionned? I'd love to have it all in the same SCSS file without tweaking the CSS.
> 
> Yes please do share your custom theme if you want help on it. Otherwise, you can get inspiration from boostwatch theme. You can see that they do have specific handling for tooltip to get the coloring right Se for example darkly
> 
> https://github.com/quarto-dev/quarto-cli/blob/88819d5b9f092381d5d20a146ac3121dc2435acc/src/resources/formats/html/bootstrap/themes/darkly.scss#L270-L273
> 
> If you have a custom them no inheriting from an existing dark theme you may want to do the same.
> 
> > Thanks for your answer! The first code block worked inside the SCSS file with no issues.
> 
> But this is basically the CSS shared above that you are happy with. It is just a matter of knowing if you want to reuse a sass variable or use specific color for your theme.

I actually checked the SCSS variables and rules files in the `quarto-cli` repo, only finding styling for the copy code button so I'm surprised that it does indeed mention the tooltip in the themes but not in the previously mentionned files.

Can you also explain what this tooltip is exactly? And why does a vague `--bg-tooltip-color` variable only change the text of the popup and not something else that could be part of this tooltip? 
And thanks for your willingness to help and further explain the matter.

**Reply by cderv** [reply]
> Can you also explain what this tooltip is exactly?

This is tooltip component from Bootstrap: https://getbootstrap.com/docs/5.3/components/tooltips/

> And why does a vague --bg-tooltip-color variable only change the text of the popup and not something else that could be part of this tooltip?

`--bg-tooltip-color` is set by Bootstrap as part of its tooltip feature
https://github.com/quarto-dev/quarto-cli/blob/88819d5b9f092381d5d20a146ac3121dc2435acc/src/resources/formats/html/bootstrap/dist/scss/_tooltip.scss#L10
and it is used by bootstrap for its tooltip element
https://github.com/quarto-dev/quarto-cli/blob/88819d5b9f092381d5d20a146ac3121dc2435acc/src/resources/formats/html/bootstrap/dist/scss/_tooltip.scss#L115
`.tooltip-inner` is what is used for the inner element of a tooltip

This is all bootstrap feature that Quarto leverage for the copy button
https://github.com/quarto-dev/quarto-cli/blob/88819d5b9f092381d5d20a146ac3121dc2435acc/src/resources/formats/html/templates/quarto-html.ejs#L248-L251

Hope it helps understand
